lib/core/src/Cardano/Wallet.hs

```diff
@@ -1589,32 +1572,49 @@ balanceTransaction
     -- doing such a thing is considered bonkers and this is not a behavior we
     -- ought to support.
 
+    let unsafeFromLovelace (Cardano.Lovelace l) = Coin.unsafeFromIntegral l
     candidateTx <- assembleTransaction $ TxUpdate
         { extraInputs
         , extraCollateral
         , extraOutputs
-        , feeUpdate = UseNewTxFee delta
+        , feeUpdate = UseNewTxFee $ unsafeFromLovelace minfee0
         }
-    let candidateMinFee = fromMaybe (Coin 0) $
-            evaluateMinimumFee tl nodePParams candidateTx
 
+    (balance, candidateMinFee) <- balanceAfterSettingMinFee candidateTx
+    surplus <- case Cardano.selectLovelace balance of
+            (Cardano.Lovelace c) | c >= 0
+                -> pure $ Coin.unsafeFromIntegral c
+                                 | otherwise
+                -> throwE
+                    . ErrBalanceTxNotYetSupported
+                    $ UnderestimatedFee
+                        (Coin.unsafeFromIntegral (-c))
+                        candidateTx
 
-    -- Fee minimization... Ideally we should factor this out and test
-    -- separately... Although, we don't want to lose the distinction between
-    -- extra outputs and normal outputs
 
-    surplus <- ExceptT . pure $ maybe
-        (Left $ ErrBalanceTxNotYetSupported $ UnderestimatedFee
-             (candidateMinFee `Coin.difference` delta)
-             candidateTx)
-        Right
-        (delta `Coin.subtract` candidateMinFee)
+    -- If there are no change outputs, "burn" the surplus as extra fee.
+    --
+    -- This should only happen when coin-selection cannot afford to construct a
+    -- change output with the appropriate minUTxOValue.
+    let extraFeeToBurn =
+            case extraOutputs of
+                [] | surplus > Coin 20_000_000 -> error $ unwords
```

Comment:
Define in terms of `minUTxOValue`

Reply:
This is not entirely straight-forward. The `minUTxOValue` isn't a protocol parameter anymore in Alonzo. 

One could perhaps use this function https://github.com/input-output-hk/cardano-wallet/blob/8106b842c80c164664030da879308ef018c33f9a/lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs#L891 if we pass in a `TokenMap.empty`, but without knowing for sure whether that is reliable, I'd feel the hard-coded 20 ada safety limit is preferable.